agent/consul/streaming_endpoint.go

```diff
@@ -28,210 +25,154 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
-			if err := sendEvent(event, aclFilter, eval, sent, server); err != nil {
-				return err
-			}
-			if event.GetEndOfSnapshot() {
-				snapshotIndex = event.Index
-			} else {
-				nSnapEvents++
-			}
-		}
-
-	} else {
-		// If there wasn't a snapshot, just send an end of snapshot message
-		// so the client knows not to wait for one.
-		endSnapshotEvent := stream.Event{
-			Topic:   req.Topic,
-			Index:   lastSentIndex,
-			Payload: &stream.Event_EndOfSnapshot{EndOfSnapshot: true},
-		}
-		snapshotIndex = lastSentIndex
-		if err := server.Send(&endSnapshotEvent); err != nil {
-			return err
-		}
-	}
-
-	h.srv.logger.Printf("[DEBUG] consul: subscribe snapshot complete snapIndex=%d "+
-		"nEvents=%d streamID=%s", snapshotIndex, nSnapEvents, streamID)
 
 	// Register a subscription on this topic/key with the FSM.
-	eventCh, err := state.Subscribe(req)
+	sub, err := state.Subscribe(req)
 	if err != nil {
 		return err
 	}
 	defer state.Unsubscribe(req)
 
-	var sentEvents uint64
+	// Deliver the events
 	for {
-		select {
-		case <-server.Context().Done():
-			return nil
-		case event, ok := <-eventCh:
-			// If the channel was closed, that means the state store filled it up
-			// faster than we could pull events out.
-			if !ok {
-				return fmt.Errorf("handler could not keep up with events")
+		events, err := sub.Next(server.Context())
+		if err == stream.ErrSubscriptionReload {
+			event := stream.Event{
+				Payload: &stream.Event_ReloadStream{ReloadStream: true},
 			}
-
-			// If we need to reload the stream (because our ACL token was updated)
-			// then pass the event along to the client and exit.
-			if event.GetReloadStream() {
-				if err := server.Send(&event); err != nil {
-					return err
-				}
-				return nil
-			}
-
-			if err := sendEvent(event, aclFilter, eval, sent, server); err != nil {
+			if err := server.Send(&event); err != nil {
 				return err
 			}
-			sentEvents++
-
-			h.srv.logger.Printf("[DEBUG] consul: subscribe sent event eventIndex=%d "+
-				"streamID=%s", event.Index, streamID)
+			h.srv.logger.Printf("[DEBUG] consul: subscribe stream reloaded "+
+				"streamID=%s", streamID)
+			return nil
 		}
-	}
-}
-
-// sendEvent sends the given event along the stream if it passes ACL, boolean, and
-// any topic-specific filtering.
-func sendEvent(event stream.Event, aclFilter *aclFilter, eval *bexpr.Evaluator,
-	sent map[uint32]struct{}, server stream.Consul_SubscribeServer) error {
-	// If it's a special case event, skip the filtering and just send it.
-	if event.GetEndOfSnapshot() || event.GetReloadStream() {
-		return server.Send(&event)
-	}
-	allowEvent := true
-
-	// Filter by ACL rules.
-	if !aclFilter.allowEvent(event) {
-		allowEvent = false
-	}
-
-	// Apply the user's boolean expression filtering.
-	if eval != nil {
-		allow, err := eval.Evaluate(event.FilterObject())
 		if err != nil {
 			return err
 		}
-		if !allow {
-			allowEvent = false
+
+		filteredEvents, err := aclFilterEvents(events, aclFilter)
```

Comment:
Interesting, I didn't know that was an option, but I don't think it "solves" the token expiration problem in the sense that we need to actively detect that and reset the stream so the client sees the permission error on a retry rather than just silently stops getting updates but will still serve the old values to the user that they aren't really allowed to any more via the agent cache.

I think token expiration is relatively easy though - we just need to create a `time.After` that forces a reset on the subscription at the expiry time.